File: Bisq/src/main/java/com/joachimneumann/bisq/CryptoHelper.java
package com.joachimneumann.bisq;

import android.util.Base64;
import java.security.NoSuchAlgorithmException;
import javax.crypto.Cipher;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

public class CryptoHelper {

    private IvParameterSpec ivspec;
    private SecretKeySpec keyspec;
    private Cipher cipher;
    private String key; // 32 character key - exchanged with phones that receive the message

    public CryptoHelper(String key_) {
        key = key_;

        keyspec = new SecretKeySpec(key.getBytes(), "AES");

        try {
            cipher = Cipher.getInstance("AES/CBC/NOPadding");
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        } catch (NoSuchPaddingException e) {
            e.printStackTrace();
        }
    }

    public String encrypt(String valueToEncrypt, String iv) throws Exception {
        while (valueToEncrypt.length() % 16 != 0) { valueToEncrypt = valueToEncrypt + " "; }

        if ( iv.length() != 16) { throw new Exception( "iv not 16 characters"); }
        ivspec = new IvParameterSpec(iv.getBytes());
        byte[] encryptedBytes = encryptInternal(valueToEncrypt, ivspec);
        byte[] encryptedBase64 = Base64.encode(encryptedBytes, Base64.DEFAULT);
        String encryptedBase64String = new String(encryptedBase64, "UTF-8");
        return encryptedBase64String;
    }

    public String decrypt(String valueToDecrypt, String iv) throws Exception {
        if ( iv.length() != 16) { throw new Exception( "iv not 16 characters"); }
        ivspec = new IvParameterSpec(iv.getBytes());
        byte[] decryptedBytes = decryptInternal(valueToDecrypt, ivspec);
        String decryptedString = new String(decryptedBytes);
        return decryptedString;
    }

    private byte[] encryptInternal(String text, IvParameterSpec ivspec) throws Exception {
        if (text == null || text.length() == 0) {
            throw new Exception("Empty string");
        }

        if (key.length() != 32) { throw new Exception("key not 32 characters"); }

        byte[] encrypted = null;
        try {
            cipher.init(Cipher.ENCRYPT_MODE, keyspec, ivspec);
            encrypted = cipher.doFinal(text.getBytes());
        } catch (Exception e) {
            throw new Exception("[encrypt] " + e.getMessage());
        }
        return encrypted;
    }

    private byte[] decryptInternal(String codeBase64, IvParameterSpec ivspec) throws Exception {
        if (codeBase64 == null || codeBase64.length() == 0) {
            throw new Exception("Empty string");
        }

        if (key.length() != 32) { throw new Exception("key not 32 characters"); }

        byte[] decrypted = null;
        try {
            cipher.init(Cipher.DECRYPT_MODE, keyspec, ivspec);
            byte[] code = Base64.decode(codeBase64, Base64.DEFAULT);
            decrypted = cipher.doFinal(code);
        } catch (Exception e) {
            throw new Exception("[decrypt] " + e.getMessage());
        }
        return decrypted;
    }
}

File: Bisq/src/main/java/com/joachimneumann/bisq/CryptoHelper.kt
package com.joachimneumann.bisq

import android.util.Base64
import java.nio.charset.Charset
import java.security.NoSuchAlgorithmException
import javax.crypto.Cipher
import javax.crypto.NoSuchPaddingException
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

class CryptoHelper(private val key: String // 32 character key - exchanged with phones that receive the message
) {

    private var ivspec: IvParameterSpec? = null
    private val keyspec: SecretKeySpec
    private var cipher: Cipher? = null

    init {

        keyspec = SecretKeySpec(key.toByteArray(), "AES")

        try {
            cipher = Cipher.getInstance("AES/CBC/NOPadding")
        } catch (e: NoSuchAlgorithmException) {
            e.printStackTrace()
        } catch (e: NoSuchPaddingException) {
            e.printStackTrace()
        }

    }

    @Throws(Exception::class)
    fun encrypt(valueToEncrypt: String, iv: String): String {
        var v = valueToEncrypt
        while (v.length % 16 != 0) {
            v = "$v "
        }

        if (iv.length != 16) {
            throw Exception("iv not 16 characters")
        }
        ivspec = IvParameterSpec(iv.toByteArray())
        val encryptedBytes = encryptInternal(v, ivspec!!)
        val encryptedBase64 = Base64.encode(encryptedBytes, Base64.DEFAULT)
        return String(encryptedBase64, Charset.forName("UTF-8"))
    }

    @Throws(Exception::class)
    fun decrypt(valueToDecrypt: String, iv: String): String {
        if (iv.length != 16) {
            throw Exception("iv not 16 characters")
        }
        ivspec = IvParameterSpec(iv.toByteArray())
        val decryptedBytes = decryptInternal(valueToDecrypt, ivspec!!)
        return String(decryptedBytes!!)
    }

    @Throws(Exception::class)
    private fun encryptInternal(text: String?, ivspec: IvParameterSpec): ByteArray? {
        if (text == null || text.length == 0) {
            throw Exception("Empty string")
        }

        if (key.length != 32) {
            throw Exception("key not 32 characters")
        }

        var encrypted: ByteArray?
        try {
            cipher!!.init(Cipher.ENCRYPT_MODE, keyspec, ivspec)
            encrypted = cipher!!.doFinal(text.toByteArray())
        } catch (e: Exception) {
            throw Exception("[encrypt] " + e.message)
        }

        return encrypted
    }

    @Throws(Exception::class)
    private fun decryptInternal(codeBase64: String?, ivspec: IvParameterSpec): ByteArray? {
        if (codeBase64 == null || codeBase64.length == 0) {
            throw Exception("Empty string")
        }

        if (key.length != 32) {
            throw Exception("key not 32 characters")
        }

        var decrypted: ByteArray?
        try {
            cipher!!.init(Cipher.DECRYPT_MODE, keyspec, ivspec)
            val code = Base64.decode(codeBase64, Base64.DEFAULT)
            decrypted = cipher!!.doFinal(code)
        } catch (e: Exception) {
            throw Exception("[decrypt] " + e.message)
        }

        return decrypted
    }
}